servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,547 @@
+import static io.servicetalk.concurrent.internal.SubscriberUtils.trySetTerminal;
+import static io.servicetalk.concurrent.internal.TerminalNotification.complete;
+import static io.servicetalk.concurrent.internal.ThrowableUtils.catchUnexpected;
+import static io.servicetalk.utils.internal.PlatformDependent.newUnboundedMpscQueue;
+import static io.servicetalk.utils.internal.PlatformDependent.throwException;
+import static java.lang.Math.min;
+import static java.util.Collections.newSetFromMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+
+final class PublisherFlatMapMerge<T, R> extends AbstractAsynchronousPublisherOperator<T, R> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(PublisherFlatMapMerge.class);
+    private final Function<? super T, ? extends Publisher<? extends R>> mapper;
+    private final int maxConcurrency;
+    private final int maxMappedDemand;
+    private final boolean delayError;
+
+    PublisherFlatMapMerge(Publisher<T> original, Function<? super T, ? extends Publisher<? extends R>> mapper,
+                          boolean delayError, Executor executor) {
+        this(original, mapper, delayError, 16, 64, executor);
+    }
+
+    PublisherFlatMapMerge(Publisher<T> original, Function<? super T, ? extends Publisher<? extends R>> mapper,
+                          boolean delayError, int maxConcurrency, int maxMappedDemand, Executor executor) {
+        super(original, executor);
+        this.mapper = requireNonNull(mapper);
+        if (maxConcurrency <= 0) {
+            throw new IllegalArgumentException("maxConcurrency: " + maxConcurrency + " (expected >0)");
+        }
+        if (maxMappedDemand <= 0) {
+            throw new IllegalArgumentException("maxMappedDemand: " + maxMappedDemand + " (expected >0)");
+        }
+        this.maxMappedDemand = maxMappedDemand;
+        this.maxConcurrency = maxConcurrency;
+        this.delayError = delayError;
+    }
+
+    @Override
+    public Subscriber<? super T> apply(final Subscriber<? super R> subscriber) {
+        return new FlatMapSubscriber<>(this, subscriber);
+    }
+
+    private static final class FlatMapSubscriber<T, R> implements Subscriber<T>, Subscription {
+        private static final Object MAPPED_SOURCE_COMPLETE = new Object();
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<FlatMapSubscriber, CompositeException> delayedErrorUpdater =
+                newUpdater(FlatMapSubscriber.class, CompositeException.class, "delayedError");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<FlatMapSubscriber> emittingLockUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(FlatMapSubscriber.class, "emittingLock");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<FlatMapSubscriber> pendingDemandUpdater =
```

Comment:
This operator is different from `flatMapMergeSingle` in that the demand isn't tightly coupled between mapped async sources and demand from downstream. The operator's demand on async sources is managed independently from downstream demand and limited by `maxMappedDemand`, at the cost of queuing internally. A previous implementation we discussed offline was distributing downstream demand to mapped sources (to avoid the queuing) but this requires a heuristic to distribute demand evenly and may lead to "dead locks" if some mapped sources don't deliver signals in a timely manner.

Reply:
the approach has changed due to offline discussions. ptal.